Create and publish typescript package

mart3323 opened the pull request:
From #todo/jsmacros/14
> 14. if someone wants to figure out the best way to repackage the typescript docs for NPM publishing, that'd be great 🙂

## Changes
### For the npm package
1. Added README.md and package.json for the NPM package
2. Removed version number from JsMacros.d.ts filename
1. Added new job in betabuild.yml to publish 0.0.0-sha (f.ex. 0.0.0-d3adb33)
1. Added new job in release.yml to publish a.b.c (f.ex. 1.7.0)
### Unrelated
1. Fixed typing of Java.from()

## Note on compatibility with main-1.19
The current 1.19 branch reports errors for basic scripts such as `Chat.log("string")` not because the packaging is wrong, but because the generated .d.ts file now contains the type `_javatypes.java.lang.Object` in a bunch of places.
For testing i put these changes directly on the 1.7.0 release and the packages produced from *that* version worked fine

mart3323 commented:
## Why publish to github and not npm
### Pros
1. It's easier, since you don't need to create an npm account and stuff the credentials for that in your secrets
2. Publishing a package on npm permanently reserves that name+version combination, so i don't feel comfortable doing test deployments there

### Cons
1. Harder to install for the user - user needs to create a github account, generate an access token, and put that in .npmrc

wagyourtail commented:
> Publishing a package on npm permanently reserves that name+version combination, so i don't feel comfortable doing test deployments there

test publishing to `@yourusername/JsMacrosAC`, and then I can change it when merging

mart3323 commented:
npm package names must be lowercase. Would `/jsmacros-types` be acceptable then?

mart3323 commented:
Updated branch `typescript-package` to latest main-1.19 (as of about half an hour ago)
Switched deployment to npm registry (`buildrelease.yml` tagged as `latest` by default, `buildbeta.yml` tagged as `beta`)
Currently published test versions are

[@mart3323/jsmacros on npm](https://www.npmjs.com/package/@mart3323/jsmacros-types)  
`@mart3323/jsmacros-types@latest` This branch, based on latest 1.19, bad types  
`@mart3323/jsmacros-types@beta` This branch rebased onto 1.7.0, good types

wagyourtail commented on .github/workflows/betabuild.yml:
```diff
@@ -10,39 +10,73 @@ on: push
+          name: ${{ steps.fname.outputs.result }}
+          path: ./dist/**/*
+
+  publish_npm:
```
beta shouldn't publish

mart3323 replied:
Will implement as requested in about 8 hours when i get home, but i would urge you to reconsider.

npm installs @latest by default (and the CI file only tags release builds with @latest), so the only way for someone to get a beta build is if they explicitly ask for it by installing either `@beta` or `0.0.0-shortsha`.
Therefore i see no downside to having both. 

On the other hand, only having release versions on npm means that anyone trying a beta (or instructed to try a beta) has to learn to do it manually anyway, at which point there's no reason for them to use the package at all since doing it manually works for *any* version

---
In the test package both tags @latest and @beta point to a beta version, that was done manually to make it easier to compare the type issues and should not happen with the CI as it currently stands (note the `--tag beta` in the betabuild publish script).

mart3323 replied:
reverted as requested

wagyourtail replied:
you do have a point... hmm...

mart3323 replied:
Restored per https://github.com/JsMacros/JsMacros/pull/62#issuecomment-1319527005

wagyourtail commented on buildSrc/src/main/java/xyz/wagyourtail/doclet/tsdoclet/Main.java:
```diff
@@ -61,7 +61,7 @@ public boolean run(DocletEnvironment environment) {
         Set<EventParser> eventClasses = new LinkedHashSet<>();
 
 
-        outputTS = new FileHandler(new File(OutputDirectory.outputDir, "JsMacros-" + Version.version + ".d.ts"));
```
is this change actually necessary? I'd prefer not to, in order to allow users of the non-npm environment to know what version their types are for

mart3323 replied:
Just seemed unnecessary for npm, since the version is in the package itself.

TODO: Put version number back, ensure tsconfig/package.json reference it correctly

mart3323 replied:
looks like i forgot to leave a comment here - done in [e4c5229](https://github.com/JsMacros/JsMacros/pull/62/commits/e4c52298dbf958a77b5a6bde2d1d94490f6ee2ac)

wagyourtail commented on docs/typescript/package.json:
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "@mart3323/jsmacros-types",
```
I have grabbed the `@jsmacros` org. want me to add you to it?

mart3323 replied:
Up to you - though keep in mind that as soon as we switch to the final package name, anything published (including from this unmerged MR) will become locked in. NPM does not allow editing/republishing to the same combination of package name and version number
source: https://docs.npmjs.com/policies/unpublish

mart3323 commented:
If/When this is otherwise ready to merge, i would like to clean up the history with a quick interactive rebase first

wagyourtail commented:
yeah. I think the requested changes are wrapping up, feel free to add back the beta actually...

I haven't really done too much with npm, but I should've invited you to the org.

mart3323 commented:
Rebase complete. publish for beta reincluded

Added one final extra change - I noticed a comment in the build script was outdated
> \# This workflow contains a single job called "build"  

so i deleted that. Then i saw all the other comments also seem like they're just from an example file.
Since none of the other .yml files had kept those, i figured i'd remove them from betabuild.yml as well to declutter it slightly

mart3323 commented:
in #72, the only feature i spotted that is missing here is building a .tgz artifact. It feels unnecessary here, since
a) You can download the package *from* npm
b) If you're not using npm and just want to drop in the types, the tsconfig + header files work perfectly

but i also don't have experience with tgz files so I might just not be thinking of the right use cases
@Johannes7k75 If we publish to npm anyway, would there still be a reason to create a tgz for the artifacts as well?

Johannes7k75 commented:
The reason i include the package in the artifact is that i dont have to get the sha of the beta build.

Johannes7k75 commented:
But other than downloading the types as a .tgz and downloading the Mod at the same time, there is no other benefit
